### ISSUE TITLE
Error: Not enough or too many segments

### ISSUE THREAD

**Issue by stefandesu** [milimetric/passport-mediawiki-oauth#2]
Hi,

we're trying to use this package for MediaWiki OAuth authentication, but the callback causes the following error:

```
Error: Not enough or too many segments
    at /srv/cocoda/login-server/node_modules/passport-mediawiki-oauth/lib/passport-mediawiki-oauth/oauth.js:173:22
    at passBackControl (/srv/cocoda/login-server/node_modules/passport-mediawiki-oauth/lib/passport-mediawiki-oauth/oauth-patches.js:114:15)
    at IncomingMessage.<anonymous> (/srv/cocoda/login-server/node_modules/passport-mediawiki-oauth/lib/passport-mediawiki-oauth/oauth-patches.js:134:13)
    at IncomingMessage.emit (events.js:194:15)
    at IncomingMessage.EventEmitter.emit (domain.js:441:20)
    at endReadableNT (_stream_readable.js:1103:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

Any ideas what we might be doing wrong? We already moved from the npm package to installing from GitHub directly, but the token time limit changes didn't seem to help.

Thanks in advance!

**Comment by milimetric** [comment]
Thanks for opening, I have been ignoring this repository for a bit.  I see some hope in new issues and people mentioning that the underlying library finally saw some love.  So I'll try to get to this, but PRs are welcome.

**Comment by stefandesu** [comment]
Thanks for your reply! I haven't looked further into this yet, but if I'm able to find out the issue, I'll try to create a PR. Any possible hints are welcome. :)

**Comment by stefandesu** [comment]
For testing purposes, I set up a [MediaWiki-Vagrant](https://www.mediawiki.org/wiki/MediaWiki-Vagrant) environment with OAuth enabled, and I'm failing to reproduce this error with the testing instance (meaning that it works fine there). This makes it much harder to debug. My suspicion though is that because the error is thrown [here](https://github.com/milimetric/passport-mediawiki-oauth/blob/master/lib/passport-mediawiki-oauth/oauth.js#L173), it is related to decoding the JWT [here](https://github.com/milimetric/passport-mediawiki-oauth/blob/master/lib/passport-mediawiki-oauth/oauth.js#L125). There isn't really any other place that could cause the error. Maybe the API returns an invalid JWT for some reason? Maybe this helps.

Edit: The OAuth application is set up in the testing environment exactly as it is in the meta wiki (apart from the callback URL of course).

**Comment by stefandesu** [comment]
I experimented a little bit further and found out that the body of the POST request to `` (which is being decoded as a JWT) is the following in my case: 

```
{"error":"mwoauth-invalid-authorization-wrong-wiki","message":"The authorization headers in your request are not valid for metawiki"}
```

Through some googling I found [this issue](https://phabricator.wikimedia.org/T226163#5272320) saying:

> So the problem is that if you select a specific project (in my case, It was common.wikimedia.org) instead of * (All projects) in OAuthConsumerRegistration. then you have to handshake with that project instead of CentralWiki.

In fact, we specified Wikidata in our registration as we are planning to allow edits specifically in Wikidata from our project. So I added the `baseURL` "https://www.wikidata.org/" to the strategy options and now it works! 

I'm closing this issue in favor of a new one which should clarify errors coming from the API.
